### ISSUE TITLE
Tracking: Diagnostic Server and EventPipe C++ to C backport

### ISSUE THREAD

**Issue by lambdageek** [dotnet/runtime#36820]
Once #41872 is merged, we will have the original C++ CoreCLR EventPipe + Diagnostic Server implementation, as well as the C shared (currently MonoVM) EventPipe Diagnostic Server code.  They should not functionally diverge.

This issue is meant to track changes that land in the CoreCLR version that should be ported over to the shared version.

So far these are:

CoreCLR PR | CoreCLR PR Status | Shared PR | Notes
-------------| --------------------|-----------|-----
#36242       | Merged                 |   -           | #37756
#36720       | Merged                 |   -           | #41152
#36733       | Merged                 |   -           | Ported as part of implementing CoreClr EventPipe shim.
#37035       | Merged                 |   -           | #37756
#37002       | Merged                 |   -           | #41152
#38967       | Merged                 |   -           | #41152
#40191       | Merged                 |   -           | #41152
#40332       | Merged                 |   -           | Ported as part of implementing CoreClr EventPipe shim.
#40499       | Merged                 |   -           | #41152
#40918       | Merged                 |   -           | Not affecting shared library.
#41052       | Merged                 |   -           | Ported as part of implementing CoreClr EventPipe shim.
#41194       | Merged                 |   -           | Not affecting shared library. CoreClr EventPipe shim includes change.
#42307       | Merged                 |   -           | Fixed in C library when initially detected and reported.
#43711       | Merged                 |   -           | #46214
#38225       | Merged                 |   -           | #46219
#44068       | Merged                 |   -           | C library already handle this scenario.
#45672       | Merged                 |   -           | #46220

**Comment by lambdageek** [comment]
@lateralusX I think you mentioned that you took the native code changes from #36242 (adding array types) already as part of #34600? Is that accurate?

**Comment by lateralusX** [comment]
Just double checked and the changes are applied but in upcoming PR, so not yet merged.

**Comment by lateralusX** [comment]
Updated table with new PR's affecting EventPipe code that needs to be ported over to Mono.

**Comment by srxqds** [comment]
could EventPipe work in mono runtime right now, it can use perf view for profiling ?

**Comment by lateralusX** [comment]
No the complete end to end is WIP, working on diagnostic server port at the moment, once that is in place we should be able to start use tools like dotnet-counters. The complete EventPipe story on Mono is a bigger work item so before we have full perf view profiling in place there are some other work items to implement as well, but once we reached first "milestone" getting dotnet-coutners and several of the tests suites running, we will lay out work items to complete the full story.

**Comment by lateralusX** [comment]
Update with new PR, #42307, affecting EventPipe/DiagnosticServer code that needs to be ported over to Mono.

**Comment by lateralusX** [comment]
Added #43711 to list of tracked PR's.

**Comment by am11** [comment]
Has there been a consideration to deduplicate the effort, by drawing out some abstraction? CoreCLR, as is, is building and linking C components (libunwind sources, libSystem.Globalization sources, version.c etc.), so the parts of EventPipe code that is impartial to OOP can be moved to `.c` files; without too much change in syntax -- if it helps the abstraction and DRY'ing.

**Comment by lateralusX** [comment]
The ported library will be shared by both runtimes and once all is done there will only be one version that can be shared between both CoreClr and Mono (the C version of the library). There is also a shim abstracting all runtime specific details into a small set of files implemented by each runtime adapting the library.

**Comment by lateralusX** [comment]
Looked over all changes to EventPipe/DiagnosticServer C++ library since last checkpoint (5:th of November), found and added #38225 #44068 #45672 to the list that needs to be evaluated for backport into C library.

**Comment by lateralusX** [comment]
Backported all PR detected so far into C library. @josalem are you aware of any PR's currently missing from the list that needs to be backported into C library? I have looked through changes since last checkpoint but couldn't detect any additional PR's than the once already added to the list.

**Comment by josalem** [comment]
I have a couple brewing, but since we've merged the C version, I can put them in directly.  I can't think of any that you haven't already handled.

**Comment by josalem** [comment]
@lateralusX now that we've switched to the C implementation by default, do we need this issue anymore?

**Comment by lateralusX** [comment]
No, since changes will now be done in the shared code, we won't need this. We still might need a way to communicate changes done in runtime specific shim implementation that might make sense to the other runtimes, but that can probably be notified per PR.
